### PR TITLE
Balder/factor out split join rebased

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli or @geirst PR
This is a rebase of #14929

### CODE REVIEW

File: storage/src/vespa/storage/persistence/testandsethelper.cpp [hunk 1]
```diff
@@ -35,7 +35,7 @@ void TestAndSetHelper::parseDocumentSelection(const document::DocumentTypeRepo &
 }
 
 spi::GetResult TestAndSetHelper::retrieveDocument(const document::FieldSet & fieldSet, spi::Context & context) {
-    return _spi.get(_env.getBucket(_docId, _cmd.getBucket()), fieldSet,_cmd.getDocumentId(),context);
+    return _spi.get(_env.getBucket(_docId, _cmd.getBucket()), fieldSet, _cmd.getDocumentId(),context);
```
Comment (vekterli) [inline storage/src/vespa/storage/persistence/testandsethelper.cpp:38]: Nit: missing space before `context`

**Review by vekterli — APPROVED** [review]
:+1: I assume the bodies of the moved handler functions are effectively unchanged? I assume as much and therefore haven't done a comprehensive diff of these.

File: storage/src/vespa/storage/persistence/asynchandler.h [hunk 2]
```diff
@@ -13,8 +13,11 @@ namespace spi {
 }
 struct PersistenceUtil;
 
+/**
+ * Handle async operations that uses a sequenced executor
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/asynchandler.h:17]: Nitpick: Missing '.' at the end. On the next sentence as well.

File: storage/src/vespa/storage/persistence/splitjoinhandler.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace storage {
+
+namespace spi { struct PersistenceProvider; }
+struct PersistenceUtil;
+class BucketOwnershipNotifier;
+class RecheckBucketInfoCommand;
+
+/**
+ * Handle operations that uses changes bucket ownership operations
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/splitjoinhandler.h:17]: There is something wrong with this sentence.
Nitpick: Missing '.' at the end. On the next sentence as well.

File: storage/src/vespa/storage/persistence/mergehandler.cpp [hunk 4]
```diff
@@ -426,6 +426,7 @@ MergeHandler::fetchLocalData(
     }
 
     document::BucketIdFactory idFactory;
+    const std::shared_ptr<const document::DocumentTypeRepo> repo = _env._component.getTypeRepo()->documentTypeRepo;
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/mergehandler.cpp:424]: Consider using `const auto`.

File: storage/src/vespa/storage/persistence/testandsethelper.cpp [hunk 5]
```diff
@@ -47,8 +47,9 @@ TestAndSetHelper::TestAndSetHelper(const PersistenceUtil & env, const spi::Persi
       _docTypePtr(_cmd.getDocumentType()),
       _missingDocumentImpliesMatch(missingDocumentImpliesMatch)
 {
-    resolveDocumentType(*env._repo);
-    parseDocumentSelection(*env._repo);
+    const std::shared_ptr<const document::DocumentTypeRepo> _repo = _env._component.getTypeRepo()->documentTypeRepo;
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/testandsethelper.cpp:50]: Consider using `const auto`.

File: storage/src/vespa/storage/persistence/simplemessagehandler.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+#include "simplemessagehandler.h"
+#include "persistenceutil.h"
+#include <vespa/persistence/spi/persistenceprovider.h>
+#include <vespa/storage/common/bucketoperationlogger.h>
+#include <vespa/document/base/exceptions.h>
+#include <vespa/document/fieldset/fieldsetrepo.h>
+
+#include <vespa/log/log.h>
+LOG_SETUP(".persistence.processall");
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/simplemessagehandler.cpp:11]: Consider updating log target.

**Review by geirst — APPROVED** [review]
👍

File: storage/src/vespa/storage/persistence/persistenceutil.cpp [hunk 7]
```diff
@@ -157,26 +155,23 @@ MessageTracker::generateReply(api::StorageCommand& cmd)
...
-      _repo(component.getTypeRepo()->documentTypeRepo),
       _spi(provider)
 {
 }
 
 PersistenceUtil::~PersistenceUtil() = default;
 
 void
-PersistenceUtil::updateBucketDatabase(const document::Bucket &bucket, const api::BucketInfo& i)
+PersistenceUtil::updateBucketDatabase(const document::Bucket &bucket, const api::BucketInfo& i) const
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistenceutil.cpp:174]: :)

File: storage/src/vespa/storage/persistence/simplemessagehandler.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+#include "messages.h"
+#include <vespa/storage/common/bucketmessages.h>
+#include <vespa/storageapi/message/persistence.h>
+
+namespace storage {
+
+namespace spi { struct PersistenceProvider; }
+struct PersistenceUtil;
+
+class SimpleMessageHandler : public Types {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/simplemessagehandler.h:15]: Fixed

File: storage/src/vespa/storage/persistence/testandsethelper.cpp (hunk 1 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/testandsethelper.cpp:38]: Fixed

File: storage/src/vespa/storage/persistence/asynchandler.h (hunk 2 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/asynchandler.h:17]: Fixed

File: storage/src/vespa/storage/persistence/splitjoinhandler.h (hunk 3 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/splitjoinhandler.h:17]: Fixed

File: storage/src/vespa/storage/persistence/mergehandler.cpp (hunk 4 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/mergehandler.cpp:424]: Fixed

File: storage/src/vespa/storage/persistence/testandsethelper.cpp (hunk 5 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/testandsethelper.cpp:50]: Fixed

File: storage/src/vespa/storage/persistence/simplemessagehandler.cpp (hunk 6 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/simplemessagehandler.cpp:11]: Fixed

**Comment by baldersheim** [comment]
@vekterli You are correct all method bodies are unchanged, unless there where bits flipping in the clipboard :)

**Pull request merged**